usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_init.c

```diff
@@ -0,0 +1,1948 @@
+	boolean_t		rval = B_FALSE;
+	pqi_raid_path_request_t	rqst;
+
+	if ((dma = pqi_alloc_single(s, len)) == NULL)
+		return (B_FALSE);
+	if (build_raid_path_request(&rqst, cmd, RAID_CTLR_LUNID,
+	    len, 0) == B_FALSE)
+		goto error_out;
+
+	rqst.rp_sglist[0].sg_addr = dma->dma_addr;
+	rqst.rp_sglist[0].sg_len = dma->len_to_alloc;
+	rqst.rp_sglist[0].sg_flags = CISS_SG_LAST;
+
+	if (submit_raid_rqst_sync(s, &rqst.header, NULL) == B_FALSE)
+		goto error_out;
+
+	(void) ddi_dma_sync(dma->handle, 0, 0, DDI_DMA_SYNC_FORCPU);
+	(void) memcpy(data, dma->alloc_memory, len);
+	rval = B_TRUE;
+
+error_out:
+	pqi_free_single(s, dma);
+	return (rval);
+}
+
+static boolean_t
+report_luns_by_cmd(pqi_state_t s, int cmd, void **buf)
+{
+	void		*data		= NULL;
+	size_t		data_len	= 0;
+	size_t		new_data_len;
+	uint32_t	new_list_len	= 0;
+	uint32_t	list_len	= 0;
+	boolean_t	rval		= B_FALSE;
+
+	new_data_len = sizeof (report_lun_header_t);
+	do {
+		if (data != NULL) {
+			PQI_FREE(data, data_len);
+		}
+		data_len = new_data_len;
+		data = PQI_ZALLOC(data_len, KM_SLEEP);
+		list_len = new_list_len;
+		if (report_luns(s, cmd, data, data_len) == B_FALSE)
+			goto error_out;
+		new_list_len =
+		    ntohl(((report_lun_header_t *)data)->list_length);
+		new_data_len = sizeof (report_lun_header_t) +
+		    new_list_len;
+	} while (new_list_len > list_len);
+	rval = B_TRUE;
+
+error_out:
+	if (rval == B_FALSE) {
+		PQI_FREE(data, data_len);
+		data = NULL;
+	}
+	*buf = data;
+	return (rval);
+}
+
+static inline boolean_t
+report_phys_luns(pqi_state_t s, void **v)
+{
+	return (report_luns_by_cmd(s, CISS_REPORT_PHYS, v));
+}
+
+static inline boolean_t
+report_logical_luns(pqi_state_t s, void **v)
+{
+	return (report_luns_by_cmd(s, CISS_REPORT_LOG, v));
+}
+
+static boolean_t
+get_device_list(pqi_state_t s, report_phys_lun_extended_t **pl,
+    report_log_lun_extended_t **ll)
+{
+	report_log_lun_extended_t	*log_data;
+	report_log_lun_extended_t	*internal_log;
+	size_t				list_len;
+	size_t				data_len;
+	report_lun_header_t		header;
+
+	if (report_phys_luns(s, (void **)pl) == B_FALSE)
+		return (B_FALSE);
+
+	if (report_logical_luns(s, (void **)ll) == B_FALSE)
+		return (B_FALSE);
+
+	log_data = *ll;
+	if (log_data) {
+		list_len = ntohl(log_data->header.list_length);
+	} else {
+		(void) memset(&header, 0, sizeof (header));
+		log_data = (report_log_lun_extended_t *)&header;
+		list_len = 0;
+	}
+
+	data_len = sizeof (header) + list_len;
+	/*
+	 * Add the controller to the logical luns which is a empty device
+	 */
+	internal_log = PQI_ZALLOC(data_len +
+	    sizeof (report_log_lun_extended_entry_t), KM_SLEEP);
+	(void) memcpy(internal_log, log_data, data_len);
+	internal_log->header.list_length = htonl(list_len +
+	    sizeof (report_log_lun_extended_entry_t));
+
+	if (*ll != NULL)
+		PQI_FREE(*ll, sizeof (report_lun_header_t) +
+		    ntohl((*ll)->header.list_length));
+	*ll = internal_log;
+	return (B_TRUE);
+}
+
+/* ---- Only skip physical devices ---- */
```

Comment:
Name this `skip_physical_device` then?  Something called `skip_device` that only skips _some_ devices is confusing.